ui/src/app/views/workspace-share/component.spec.tsx

```diff
@@ -0,0 +1,140 @@
+      closeFunction: () => {},
+      workspace: tomRiddleDiary,
+      accessLevel: WorkspaceAccessLevel.OWNER,
+      userEmail: 'harry.potter@hogwarts.edu',
+      sharing: true
+    };
+  });
+
+  it('display correct users', () => {
+    const wrapper = component();
+    expect(wrapper).toBeTruthy();
+    expect(wrapper.find('[data-test-id="collab-user-name"]').length).toBe(3);
+    expect(wrapper.find('[data-test-id="collab-user-email"]').length).toBe(3);
+    const expectedNames = [harry, hermione, ron].map(u => u.givenName + ' ' + u.familyName);
+    expect(wrapper.find('[data-test-id="collab-user-name"]').map(el => el.text())).toEqual(expectedNames);
+    const expectedEmails = [harry, hermione, ron].map(u => u.email);
+    expect(wrapper.find('[data-test-id="collab-user-email"]').map(el => el.text())).toEqual(expectedEmails);
+  });
+
+  it('displays correct role info', () => {
+    const wrapper = component();
+    expect(wrapper.find(getSelectString(harryRole)).text()).toEqual(fp.capitalize(WorkspaceAccessLevel[harryRole.role]));
+    expect(wrapper.find(getSelectString(hermioneRole)).text()).toEqual(fp.capitalize(WorkspaceAccessLevel[hermioneRole.role]));
+    expect(wrapper.find(getSelectString(ronRole)).text()).toEqual(fp.capitalize(WorkspaceAccessLevel[ronRole.role]));
+    expect(wrapper.find(getSelectString(lunaRole)).length).toBe(0);
+  });
+
+  it('removes user correctly', () => {
+    const wrapper = component();
+    wrapper.find('[data-test-id="remove-collab-ron.weasley@hogwarts.edu"]').first().simulate('click');
+    const expectedNames = [harry, hermione].map(u => u.givenName + ' ' + u.familyName);
+    expect(wrapper.find('[data-test-id="collab-user-name"]').map(el => el.text())).toEqual(expectedNames);
+  });
+
+  it('adds user correctly', async() => {
+    let clock = Lolex.install({shouldAdvanceTime: true});
+    const wrapper = component();
+    wrapper.find('[data-test-id="search"]').simulate('change', {target: {value: 'luna'}});
+    clock.tick(401);
+    await waitOneTickAndUpdate(wrapper);
+    wrapper.update();
+    clock.uninstall();
+    wrapper.find('[data-test-id="add-collab-luna.lovegood@hogwarts.edu"]').first().simulate('click');
+    const expectedNames = [harry, hermione, ron, luna].map(u => u.givenName + ' ' + u.familyName);
+    expect(wrapper.find('[data-test-id="collab-user-name"]').map(el => el.text())).toEqual(expectedNames);
+  });
+
+  it('does not allow self-removal', () => {
+    const wrapper = component();
+    const dataString = '[data-test-id="remove-collab-harry.potter@hogwarts.edu"]';
+    expect(wrapper.find(dataString).length).toBe(0);
+  });
+
+  it('does not allow self role change', () => {
+    const wrapper = component();
+    expect(wrapper.find('[data-test-id="harry.potter@hogwarts.edu-user-role"]').first()
+      .props()['isDisabled']).toBe(true);
+  });
+
+  it('saves acl correctly after changes made', async() => {
+    let clock = Lolex.install({shouldAdvanceTime: true});
```

Comment:
Mostly for my understanding: any reason not to put this in setup/teardown? Are there downsides to using this in a test?

Reply:
Not specifically that I am aware of. This is just the only test where it's needed so I thought keeping it  here would make troubleshooting issues later easier.